### PR TITLE
[#2736] feat(spark-connector) support hive external table

### PR DESCRIPTION
**Pull request by charliecheng630**
### What changes were proposed in this pull request?

support hive external table format
`CREATE EXTERNAL TABLE family (id INT, name STRING)`

### Why are the changes needed?

Fix: #2736 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

UT and IT

### CODE REVIEW

File: spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/hive/TestHivePropertiesConverter.java [hunk 1]
```diff
@@ -76,6 +76,14 @@ void testTableFormat() {
                 "b"));
     Assertions.assertEquals(
         ImmutableMap.of(TableCatalog.OPTION_PREFIX + "a", "a", "b", "b"), hiveProperties);
+
+    // external table
```
Comment (FANNG1) [inline spark-connector/src/test/java/com/datastrato/gravitino/spark/connector/hive/TestHivePropertiesConverter.java:80]: please create a new `testExternalTable`

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConstants.java [hunk 2]
```diff
@@ -27,12 +29,15 @@ public class HivePropertiesConstants {
...
   public static final String GRAVITINO_HIVE_FORMAT_CSV = StorageFormat.CSV.toString();
-
+  public static final String GRAVITINO_HIVE_EXTERNAL_TABLE = EXTERNAL_TABLE.name();
+  public static final String GRAVITINO_HIVE_TABLE_TYPE = "table-type";
   public static final String SPARK_HIVE_STORED_AS = "hive.stored-as";
   public static final String SPARK_HIVE_INPUT_FORMAT = "input-format";
   public static final String SPARK_HIVE_OUTPUT_FORMAT = "output-format";
   public static final String SPARK_HIVE_SERDE_LIB = "serde-lib";
 
+  public static final String SPARK_HIVE_EXTERNAL = "external";
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConstants.java:39]: please sperate  Spark and Hive properties with new lines, and remove lines before `SPARK_HIVE_EXTERNAL`

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java [hunk 3]
```diff
@@ -201,6 +201,25 @@ void testHiveFormatWithStoredAs() {
...
+    String createTableSql = getCreateExternalTableString(tableName);
+    sql(createTableSql);
+    SparkTableInfo tableInfo = getTableInfo(tableName);
+
+    SparkTableInfoChecker checker =
+        SparkTableInfoChecker.create()
+            .withName(tableName)
+            .withTableProperties(
+                ImmutableMap.of(
+                    HivePropertiesConstants.GRAVITINO_HIVE_TABLE_TYPE,
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java:217]: you should transform Gravitino hive properties to spark properties, and check with spark properties

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java [hunk 4]
```diff
@@ -679,6 +679,12 @@ protected String getCreateSimpleTableString(String tableName) {
         tableName);
   }
 
+  protected String getCreateExternalTableString(String tableName) {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:682]: suggest to do some abstraction with `getCreateSimpleTableSql`.

Comment (charliecheng630) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:682]: How about declare an abstract method in `SparkEnvIT.java`
` protected abstract String getCreateSimpleTableString(String... args);`

and the implementation in `SparkHiveCatalogIT.java `

```
@Override
  protected String getCreateSimpleTableString(String... args) {
    if(args.length == 1) {
      return String.format(
              "CREATE TABLE %s (id INT COMMENT 'id comment', name STRING COMMENT '', age INT)",
              args[0]);
    } else {
      return String.format(
              "CREATE %s TABLE %s (id INT COMMENT 'id comment', name STRING COMMENT '', age INT)",
              args[0], args[1]);
    }
  }
```

Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:682]: defining `tableName` and `isExternal` seems more clear.
```java
  protected String getCreateSimpleTableString(String tableName) {
    return getCreateSimpleTableString(tableName, false);
  }

  protected String getCreateSimpleTableString(String tableName, boolean isExternal) {
   // generate create table SQL
 }

```

**Comment by charliecheng630** [comment]
Updated.  Please help to review this PR again, thanks! @FANNG1

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java [hunk 5]
```diff
@@ -70,6 +75,11 @@ public Map<String, String> toGravitinoTableProperties(Map<String, String> proper
       }
     }
 
+    if (isExternal.equals("true")) {
+      gravitinoTableProperties.put(
+          HivePropertiesConstants.GRAVITINO_HIVE_TABLE_TYPE,
+          HiveTablePropertiesMetadata.TableType.EXTERNAL_TABLE.name());
+    }
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java:82]: please add opposite transform logic to `toSparkTableProperties`

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java [hunk 6]
```diff
@@ -674,9 +674,19 @@ protected String getExpectedTableData(SparkTableInfo table) {
   }
 
   protected String getCreateSimpleTableString(String tableName) {
-    return String.format(
-        "CREATE TABLE %s (id INT COMMENT 'id comment', name STRING COMMENT '', age INT)",
-        tableName);
+    return getCreateSimpleTableString(tableName, false);
+  }
+
+  protected String getCreateSimpleTableString(String tableName, boolean isExternal) {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:680]: following code seems more extensible, we may add more parameters beside isExternal.
```java
String external = "";
if (isExternal) {
  external = "EXTERNAL";
}

String.format(
          "CREATE %s TABLE %s (id INT COMMENT 'id comment', name STRING COMMENT '', age INT)",
          external,tableName);

```

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java [hunk 7]
```diff
@@ -201,6 +201,23 @@ void testHiveFormatWithStoredAs() {
     checkParquetFile(tableInfo);
   }
 
+  @Test
+  void testHiveFormatWithExternalTable() {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java:205]: for external table, one of feature is the data retains after dropping table, could you add related test?
1. set table location explicitly when creating table.
2. add `dropTable` after `checkTabelReadWrite`
3. check if data file still exists in the table location, add `checkDataFileExists`   like `checkDirExists`

Comment (charliecheng630) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java:205]: Sure! I am glad to do this. :)

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java [hunk 8]
```diff
@@ -118,6 +118,10 @@ protected void dropTableIfExists(String tableName) {
     sql("DROP TABLE IF EXISTS " + tableName);
   }
 
+  protected void dropTable(String tableName) {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/util/spark/SparkUtilIT.java:121]: how about use `dropTableIfExists`?

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java [hunk 9]
```diff
@@ -598,6 +603,26 @@ protected void checkDirExists(Path dir) {
     }
   }
 
+  protected void checkDataFileExists(Path dir) {
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:606]: `checkDataFileExists` suggest just check the data file exists under dir and no need to check fileList.size and file content.

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java [hunk 10]
```diff
@@ -201,6 +201,26 @@ void testHiveFormatWithStoredAs() {
...
+    sql(createTableSql);
+    SparkTableInfo tableInfo = getTableInfo(tableName);
+
+    SparkTableInfoChecker checker =
+        SparkTableInfoChecker.create()
+            .withName(tableName)
+            .withTableProperties(
+                ImmutableMap.of(HivePropertiesConstants.SPARK_HIVE_EXTERNAL, "true"));
+    checker.check(tableInfo);
+    checkTableReadWrite(tableInfo);
```
Comment (FANNG1) [inline integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/hive/SparkHiveCatalogIT.java:218]: add a blank line after `checkTableReadWrite`

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java [hunk 11]
```diff
@@ -70,6 +75,11 @@ public Map<String, String> toGravitinoTableProperties(Map<String, String> proper
       }
     }
 
+    if (isExternal.equals("true")) {
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java:78]: equalsIgnorecase?

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java [hunk 12]
```diff
@@ -98,11 +108,21 @@ static Map<String, String> toOptionProperties(Map<String, String> properties) {
                     return TableCatalog.OPTION_PREFIX
                         + key.substring(
                             HivePropertiesConstants.GRAVITINO_HIVE_SERDE_PARAMETER_PREFIX.length());
+                  } else if (key.equals(HivePropertiesConstants.GRAVITINO_HIVE_TABLE_TYPE)) {
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java:111]: please don't mix the external logic and option process logic, it's hard to maintain.

**Comment by charliecheng630** [comment]
Updated. Please help to review this PR again, thanks! @FANNG1

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java (hunk 12 above)
Comment (FANNG1) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java:111]: @charliecheng630 could you fix this?  move `external` logic from `toOptionProperties`

File: spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java [hunk 13]
```diff
@@ -83,7 +93,17 @@ public Map<String, String> toGravitinoTableProperties(Map<String, String> proper
 
   @Override
   public Map<String, String> toSparkTableProperties(Map<String, String> properties) {
-    return toOptionProperties(properties);
+    Map<String, String> sparkTableProperties = toOptionProperties(properties);
+    String externalValue =
```
Comment (FANNG1) [inline spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java:97]: the default value of `HivePropertiesConstants.GRAVITINO_HIVE_TABLE_TYPE` is not `false`, and we shouldn't depend on it.  could you rewrite like this?
```java
String hiveTableType = sparkTableProperties.get(HivePropertiesConstants.GRAVITINO_HIVE_TABLE_TYPE)
if (HivePropertiesConstants.GRAVITINO_HIVE_EXTERNAL_TABLE.equalsIgnoreCase(hiveTableType)) {
  xxx
}
```

Comment (charliecheng630) [reply spark-connector/src/main/java/com/datastrato/gravitino/spark/connector/hive/HivePropertiesConverter.java]: Updated. Please help to review again, thanks!!!!!!!!

File: integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java (hunk 6 above)
Comment (FANNG1) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:748]: @charliecheng630 , could you fix this?

Comment (charliecheng630) [reply integration-test/src/test/java/com/datastrato/gravitino/integration/test/spark/SparkCommonIT.java:748]: Updated.

**Comment by FANNG1** [comment]
LGTM

**Review by FANNG1 — APPROVED** [review]

**Comment by FANNG1** [comment]
@charliecheng630  merged to main, thanks for your work!

**Pull request merged**